File: frontend/src/lib/queries/insights.ts
'use client'

import { useMutation } from '@tanstack/react-query'
import { useApi } from '@/lib/api/use-api'
import type { DashboardResult } from '@/lib/queries/visualization'

export type Grounding = {
  status: 'none' | 'clean' | 'partial' | 'weak'
  label: string
  total: number
  verified_count: number
  coverage: number
}

/** One computed figure the report was allowed to cite, with its derivation. */
export type Figure = {
  key: string
  label: string
  value: number
  unit: string
  display: string
  formula: string
  period: string
  quality: 'exact' | 'scenario'
  assumption: string
}

/** How every number in the report got there. */
export type Verification = {
  status: 'verified' | 'unverified'
  label: string
  figures_cited_from_engine: number
  figures_typed_by_model: number
  typed_and_verified: number
  citation_rate: number
  unverified_figures: { figure: string; value: number; context: string }[]
  unknown_citations: string[]
  unsupported_currency_claims: string[]
  misvalued_actions: string[]
  ambiguous_citations: string[]
  corrected: boolean
}

/** A finding, ranked by the money it puts at stake. */
export type EvidenceItem = {
  rank: number
  tag: 'THREAT' | 'LEAK' | 'RISK' | 'OPPORTUNITY' | 'STRENGTH' | 'CAVEAT'
  section: string
  money_at_stake: number
  text: string
}

export type InsightsResult = {
  report_md: string
  template: string
  grounding: Grounding
  analytics: Record<string, any>
  dashboard: DashboardResult
  verification?: Verification | null
  figures?: Figure[]
  evidence?: EvidenceItem[]
  blind_spots?: string[]
  decision?: Record<string, any>
}

export function useGenerateInsights(projectId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (input: { business_context: string; template_override?: string | null }) =>
      call<InsightsResult>(`/api/v1/projects/${projectId}/insights`, { method: 'POST', json: input }),
  })
}

export type ChatMessage = { role: 'user' | 'assistant'; content: string }

export type ChatResult = {
  answer: string
  output: string
  error: string | null
  figure: { data: any[]; layout: Record<string, any> } | null
  preview: Record<string, unknown>[] | null
}

export function useInsightsChat(projectId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (input: { message: string; report_md: string; template_label: string; history: ChatMessage[] }) =>
      call<ChatResult>(`/api/v1/projects/${projectId}/insights/chat`, { method: 'POST', json: input }),
  })
}

File: backend/app/schemas/insights.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel

from backend.app.schemas.visualization import DashboardResponse


class InsightsRequest(BaseModel):
    business_context: str = ""
    template_override: str | None = None
    model: str | None = None


class GroundingOut(BaseModel):
    status: str
    label: str
    total: int
    verified_count: int
    coverage: float
    # Figures the checker could not trace to a computed value (surfaced so the
    # reader can scrutinise them directly).
    unverified: list[str] = []
    # Per-figure citation: each verified number and the exact source value it
    # was matched against — the machine-checkable audit trail behind the badge.
    traces: list[dict[str, Any]] = []


class FigureOut(BaseModel):
    """One computed figure the report was allowed to cite, with its provenance.

    Surfaced so a reader can check any number in the prose against the formula
    that produced it — the audit trail is part of the product, not a debug aid.
    """

    key: str
    label: str
    value: float
    unit: str
    display: str
    formula: str
    period: str = ""
    quality: str = "exact"
    assumption: str = ""


class VerificationOut(BaseModel):
    """Verdict on how every number in the report got there."""

    status: str
    label: str
    # Figures inserted verbatim by the calculation engine (impossible to fabricate).
    figures_cited_from_engine: int = 0
    # Figures the model typed itself, and how many of those checked out.
    figures_typed_by_model: int = 0
    typed_and_verified: int = 0
    citation_rate: float = 1.0
    unverified_figures: list[dict[str, Any]] = []
    unknown_citations: list[str] = []
    # Currency symbols the dataset cannot support.
    unsupported_currency_claims: list[str] = []
    # Recommendations whose stated value cites a level rather than a change.
    misvalued_actions: list[str] = []
    # Figures written without naming the metric they belong to.
    ambiguous_citations: list[str] = []
    corrected: bool = False


class InsightsResponse(BaseModel):
    report_md: str
    template: str
    grounding: GroundingOut
    analytics: dict[str, Any]
    dashboard: DashboardResponse
    # ── Audit surface ──────────────────────────────────────────────────────
    verification: VerificationOut | None = None
    figures: list[FigureOut] = []
    # Findings ranked by money at stake — the computed priority order the
    # report was written against.
    evidence: list[dict[str, Any]] = []
    # Questions this dataset provably cannot answer.
    blind_spots: list[str] = []
    # The full deterministic decision layer (period bridges, concentration,
    # margin, discount, scenarios) for the UI to render directly.
    decision: dict[str, Any] = {}


class InsightsChatMessage(BaseModel):
    role: str
    content: str


class InsightsChatRequest(BaseModel):
    message: str
    report_md: str
    template_label: str = ""
    history: list[InsightsChatMessage] = []
    model: str | None = None


class InsightsChatResponse(BaseModel):
    answer: str
    output: str = ""
    error: str | None = None
    figure: dict[str, Any] | None = None
    preview: list[dict[str, Any]] | None = None
    grounded: bool = True
